src/scene/materials/shader-graph-builder.js

```diff
@@ -0,0 +1,128 @@
+import { ShaderGraphNode } from './shader-graph-node.js';
+
+var id = 0;
+
+/**
+ * @private
+ * @class
+ * @name pc.ShaderGraphBuilder
+ * @classdesc A Shader Graph Builder class
+ * @param {pc.Application} app - Application with shader graph core nodes registered that will be used to build the shader graph
+ */
+var ShaderGraphBuilder = function (app) {
+    id++;
+
+    this.app = app;
+    this._nodes = this.app.shaderNodes._nodes;
+
+    this._graph = new ShaderGraphNode();
+    this._graph.name = 'graphRoot_' + id;
+};
+
+ShaderGraphBuilder.prototype.constructor = ShaderGraphBuilder;
+
+Object.assign(ShaderGraphBuilder.prototype, {
+    /**
+     * @private
+     * @function
+     * @name pc.ShaderGraphBuilder#addParam
+     * @description adds a parameter input to graph
+     * @param {string} type - type of parameter
+     * @param {string} name - name of parameter
+     * @param {any} value - optional value of parameter
+     * @returns {any} returns the created input port
+     */
+    addParam: function (type, name, value) {
+        var ioPort = this._graph.addInput(type, name, value);
+        return ioPort;
+    },
+    /**
+     * @private
+     * @function
+     * @name pc.ShaderGraphBuilder#addNode
+     * @description creates and adds a core node to the shader graph and connects up inputs
+     * @param {string} name - name of parameter
+     * @param {object} [arguments] - first argument is name string, then 0-N output node ports or params to connect to new core node inputs - number and port type of arguments must match.
+     * @returns {number} returns the created node id
+     * @example
+     * var node0 = pc.shadergraph.addNode('uv0');
+     * var node1 = pc.shadergraph.addNode('texSample', param0, node0 );
+     * var node2 = pc.shadergraph.addNode('splitVec4', node1 );
+     * var node3 = pc.shadergraph.addNode('mul3', param1, node42 );
+     */
+    addNode: function (name) {
+        var args = arguments;
+        var coreNode = this._nodes[name];
+        var sgIndex = this._graph.addSubGraph(coreNode);
+
+        var sgInputs = coreNode.graphData.ioPorts.filter(function (ioPort) {
+            return ioPort.name.startsWith('IN_');
+        });
+
+        if (sgInputs.length === (args.length - 1)) {
+            for (var argIndex = 0; argIndex < (args.length - 1); argIndex++) {
+                var arg = args[argIndex + 1];
+                if (typeof(arg) === 'number') {
+                    // ret port of a node
+                    var argNodeIndex = arg;
+                    this._graph.connect(argNodeIndex, 'OUT_ret', sgIndex, sgInputs[argIndex].name);

```

Comment:
I'd like this to be more robust as it's public API - this should check .connect was successful, and print errors with details if not.

Reply:
ShaderGraphNode won't be made public - additional checks could be made in the builder when/if it gets made public (after MVP).

Reply:
added validation code to the builder class